### ISSUE TITLE
code: implement nn.Module

### ISSUE THREAD

**Issue by clay-arras** [UBCAgroBot/fruit-multiclass-detection#13]
## Prerequisites

You should already have your basic autograd engine and tensor operations implemented.
This module will build on top of that to make it easier to organize models and parameters.

## Motivation

Right now, all parameters and computations are handled manually. As we add more layers and build larger models, this becomes messy.
The goal here is to create a lightweight module system (like `torch.nn.Module`) that:

* automatically tracks parameters when they’re assigned,
* provides a simple `parameters()` method to collect them, and
* gives a consistent interface for calling and zeroing gradients.

This will make it much easier to manage models and optimizers later on.

## Overview

* `__init__()`
* `forward()`: abstract function that takes arbitrary keywords and arguments
* `__call__()` — calls `forward()` (not abstract)
* `__setattr__()` — handles attribute assignment
* `parameters()` — recursively collects parameters and yields them (use a Generator)
* `zero_grad()` — sets all parameter gradients to zero

Here's an example of what automatically registering parameters might look like:

```
class T(Module):
    def __init__(self):
        self.x = Value(...)

    def forward(self):
        ...
```

Then:

```
T().parameters()
# should yield self.x
```

Essentially, when we initialize a variable in `__init__`, we want to have it automatically registered s.t. when we call parameters(), it shows up. This should also happen for modules, i.e: 

```
class N(Module):
    def __init__(self):
        self.a = Value(...)
        self.b = T(...)
```

Then:

```
N().parameters()
# should yield x and a
```
